### PR TITLE
[SIM_FLUTTER-74] [Integrate] Implement user can withdraw cash

### PR DESCRIPTION
**Pull request by jescabillas**
### Issue links
[backlog](https://framgiaph.backlog.com/view/SIM_FLUTTER-74)
[Slack](https://sun-philippine.slack.com/archives/C059L7XAXFB/p1688553844686229)

### Definition of Done
 - [x] User will be able to withdraw through dashboard 
 - [x] User will be able to withdraw through account details 
 - [x] Display necessary messages for validation errors
 - [x] Once transaction successfully, remove the withdraw form and display modal/pop-up for success message  

### Notes
- Only in withdraw modal/pop-up form 
- Disregard the displaying account balance in dashboard since it has issues raised

### Recording

https://github.com/framgia/sph-flutter/assets/114911074/3c3830fd-8e97-42b7-8028-98229e3ea023

image

### CODE REVIEW

**Comment by Dbulawan13** [comment]
Comment:

- [x] Please update the validation to be the same as other pages, such as login, signup, and so on.
`Currently:`
- Once `Amount` and `Description` field is empty. `Confirm` button is enable and when user click a validation modal will appear.

https://github.com/framgia/sph-flutter/assets/86758474/9cb03642-0d90-4994-900c-fc45500cd3d2

`Should be:`
- `Confirm` button should be disable as default, then once user filled up the `Amount` and `Description` fields, Confirm button will be enable. Then once user clear the fields an error message will appear stating "This field cannot be empty."

https://github.com/framgia/sph-flutter/assets/86758474/c03adf9f-b469-42fb-807e-4c335dafa54d

- [ ] Starting Balance will not update everytime user withdraw.

https://github.com/framgia/sph-flutter/assets/86758474/9712a8db-44be-4412-b2e1-173036faf996

Comments: 2

File: frontend/lib/src/components/transaction_component/transaction_component.dart [hunk 1]
```diff
@@ -6,56 +9,109 @@ import 'package:get/get.dart';
 import 'package:frontend/src/components/button.dart';
 import 'package:frontend/src/components/input/input_field.dart';
 import 'package:frontend/src/helper/dialog/show_alert_dialog.dart';
+import 'package:frontend/src/helper/dio.dart';
```
Comment (rdelrosariojr) [inline frontend/lib/src/components/transaction_component/transaction_component.dart]: sort the imports

File: frontend/lib/src/components/transaction_component/transaction_component.dart [hunk 2]
```diff
@@ -6,56 +9,109 @@ import 'package:get/get.dart';
...
+        if (amountValue == null ||
+            descriptionValue == null ||
+            (type.contains('TRANSFER') && receiverValue == null)) {
+          alertDialog();
 
           return;
         }
 
-        Get.back();
+        final transactionResponse = await client.post(
```
Comment (rdelrosariojr) [inline frontend/lib/src/components/transaction_component/transaction_component.dart]: please create transaction service for this. refer to service folder.

File: frontend/lib/src/components/transaction_component/transaction_component.dart [hunk 3]
```diff
@@ -6,56 +9,109 @@ import 'package:get/get.dart';
...
+          accountTransactionsUrl.replaceFirst('{id}', accountId),
+          data: {
+            'amount': amountValue,
+            'description': descriptionValue,
+            'transaction_type': type,
```
Comment (rdelrosariojr) [inline frontend/lib/src/components/transaction_component/transaction_component.dart]: for transaction type, please use enum. you can refer to my draft PR for transfer integration, and check the transaction model.

File: frontend/lib/src/components/transaction_component/transaction_component.dart [hunk 4]
```diff
@@ -6,56 +9,109 @@ import 'package:get/get.dart';
...
+            'category': typeTexts[type]['category'],
```
Comment (rdelrosariojr) [inline frontend/lib/src/components/transaction_component/transaction_component.dart]: same with category

File: frontend/lib/src/components/transaction_component/transaction_component.dart [hunk 5]
```diff
@@ -74,10 +130,9 @@ class TransactionComponent extends StatelessWidget {
                 height: 13,
               ),
               if (type == 'TRANSFER') ...[
-                InputField(
-                  name: 'email',
-                  label: 'Email',
-                  validator: FormBuilderValidators.email(),
+                const InputField(
```
Comment (rdelrosariojr) [inline frontend/lib/src/components/transaction_component/transaction_component.dart]: don't make changes for transfer for now since it is for other task.

**Review by rdelrosariojr — COMMENTED** [review]
reviewed time: 0.6
comments: 5

**Comment by rdelrosariojr** [comment]
please also rebase to PR49 so you can dynamically change the account balance in the account page.

**Comment by jescabillas** [comment]
> Comment:
> 
> * [x]  Please update the validation to be the same as other pages, such as login, signup, and so on.
>   `Currently:`
> * Once `Amount` and `Description` field is empty. `Confirm` button is enable and when user click a validation modal will appear.
> 
>  Peek.2023-07-06.13-55.mp4 
> `Should be:`
> 
> * `Confirm` button should be disable as default, then once user filled up the `Amount` and `Description` fields, Confirm button will be enable. Then once user clear the fields an error message will appear stating "This field cannot be empty."
> 
>  reference.mp4 
> * [ ]  Starting Balance will not update everytime user withdraw.
> 
>  Peek.2023-07-06.13-58.mp4 
> Comments: 2

Applied changes based from comment above.
As for the 2nd comment above, I put notes above to note that the balance in dashboard is not working. (this is because of yesterday's bug found for getting the balance api)

File: frontend/lib/src/components/transaction_component/transaction_component.dart (hunk 1 above)
Comment (jescabillas) [reply frontend/lib/src/components/transaction_component/transaction_component.dart]: Please confirm what do you mean buy sort? It has the same pattern with the other components

File: frontend/lib/src/components/transaction_component/transaction_component.dart (hunk 3 above)
Comment (jescabillas) [reply frontend/lib/src/components/transaction_component/transaction_component.dart]: Used enum

File: frontend/lib/src/components/transaction_component/transaction_component.dart (hunk 1 above)
Comment (rdelrosariojr) [reply frontend/lib/src/components/transaction_component/transaction_component.dart]: @jescabillas sorry haha namalikmata. Just disregard this one.

File: frontend/lib/src/components/transaction_component/transaction_component.dart (hunk 4 above)
Comment (jescabillas) [reply frontend/lib/src/components/transaction_component/transaction_component.dart]: Used enum

File: frontend/lib/src/components/transaction_component/transaction_component.dart (hunk 5 above)
Comment (jescabillas) [reply frontend/lib/src/components/transaction_component/transaction_component.dart]: Reverted

File: frontend/lib/src/components/transaction_component/transaction_component.dart (hunk 2 above)
Comment (jescabillas) [reply frontend/lib/src/components/transaction_component/transaction_component.dart]: Added service provider for transaction

**Review by Dbulawan13 — APPROVED** [review]
Approve PR #50 

Working:
- Able to withdraw cash
- Once user withdraw, current balance will be deducted correctly
- Unable to withdraw if all fields are not filled up
- Unable to withdraw if user input insufficient amount
- Unable to withdraw if user input non integer

https://github.com/framgia/sph-flutter/assets/86758474/53bd94c9-4df0-4dcd-ba94-9d29fa65210b

Testing Hours: 2
Comments: 2
Iteration: 1

**Review by rdelrosariojr — APPROVED** [review]
**APPROVED PR50**

reviewed time: 0.8
comments: 5

**Pull request merged**